### ISSUE TITLE
If using a Java version earlier than Java 11, it is not possible to unload a custom class loader after executing the bootstrap method (**not a ByteBuddy issue**).

### ISSUE THREAD

**Issue by dogourd** [raphw/byte-buddy#1432]
I've implemented a javaagent and load ByteBuddy's classes and my own classes from a specific class loader(URLClassLoader). 

However, after closing this class loader and removing the explicit reference to it, the related classes still couldn't be unloaded. Eventually, I found that the issue was caused by `net.bytebuddy.description.type.TypeDescription.ForLoadedType#TYPE_CACHE` (version 1.14.3), where ByteBuddy cached `TargetType.class`. This was the root cause of the problem. 

When I cleared this map using reflection (`TYPE_CACHE.clear()`) or removed the `TargetType.class` element (`TYPE_CACHE.remove(TargetType.class)`), the class unloading could be performed normally. 

Therefore, I suggest setting it as a WeakMap to avoid strong references to Class<TargetType>, or adding a singleton object such as `ForLoadedTypeCacheHolder` and exposing TYPE_CACHE as an instance field for ForLoadedType to use. 

I've done a simple verification locally and it meets my expectations.
Like this(Just an example):
```java
class ForLoadedType extends AbstractBase implements Serializable {
    private static ForLoadedType CACHE_HOLDER = new ForLoadedType();
    // Remove the `static` modifier.
    private /*static*/ Map<Class<?>, TypeDescription> TYPE_CACHE;
    
    // remove the static block
    /*
    static {
        // ...
    }
    */

    private ForLoadedType() {
        TYPE_CACHE = new HashMap<>();
        // put cache objects.
        // ...
    }

    public static TypeDescription of(Class<?> type) {
            TypeDescription typeDescription = CACHE_HOLDER.TYPE_CACHE.get(type);
            return typeDescription == null
                    ? new ForLoadedType(type)
                    : typeDescription;
        }
}
```

@raphw Do you think this is an issue with ByteBuddy or did I miss some best practices for using ByteBuddy?

**Comment by raphw** [comment]
That surprises me. `TypeDescription` and `TargetType` should be loaded by the same class loader. How does the reference to the same class loader prevent its unloading?

**Comment by dogourd** [comment]
Yes, you're right. This was indeed unrelated to the class, as it was eventually possible to uninstall it normally. However, during the process, I made many attempts and couldn't determine why the modifications were necessary to achieve the desired result.

Now I've encountered another problem: after using the `WithCustomMapping#bootstrap` API, the classloader cannot be unloaded once the instrumented method is called. I tried exporting a heapdump file and analyzing it with JProfiler, which shows that the classloader no longer exists. However, if I attach JVisualVM to the target process, it will display that there is still an instance of the classloader in memory, but its retention size is 0 (this is also the case with the jhat tool), and the JVM never unloads this classloader.

To reproduce this problem, I created a minimal example as possible. I want to know if invoking the `invokedynamic` instruction generates some content that is not visible in the heapdump and references the custom classloader. If so, how should I handle this situation? If you have the time, please take a look [here](https://github.com/dogourd/agent-detach-issue). If you need any further information, please let me know.

**Comment by raphw** [comment]
That's an issue that cannot really be solved by Byte Buddy, though. If you link a class to your agent somehow, those classes are linked forever. Normally one would put a class on the bootloader when using it for bootstrapping.

**Comment by dogourd** [comment]
Yes, but these links should disappear when we call `removeClassFileTransformer` and retransform the classes. I'm now highly suspicious that there may be some issues with the old JVM in handling this. In the example I mentioned earlier, if we start the application with Java 8-10 versions, the class loaders cannot be unloaded after detaching the agent, but switching to Java 11 made unloading work properly. I can confirm that I have not made any code changes. I plan to try various garbage collectors and consult Java 11 release notes to see what changes have been made.

**Comment by raphw** [comment]
There's a chance that the boot methods stay cached. Did you try to find the actual leak in a heap dump or similar?

**Comment by dogourd** [comment]
Ah, yes, I did manage to obtain a heap dump, but I didn't find any strong references to the class loader (at least that's what it seems to me). I believe there are indeed no references to the class loader anymore, otherwise why would it be possible to unload it in Java 11?

I've already spent quite some time on this issue, and now I'm pretty much stuck. If you're interested in this problem, I hope you can take a look at the example repository [agent-detach-issue](https://github.com/dogourd/agent-detach-issue) I created for this issue. I've explained how to reproduce the problem there and uploaded a heap dump file. In this example project, we need to pay attention to the MyURLClassLoader class loader, which loads classes starting with net.bytebuddy. and org.example.

**Comment by raphw** [comment]
Could you try setting `net.bytebuddy.nexus.disabled` to `false` as a system property?

**Comment by dogourd** [comment]
> Could you try setting net.bytebuddy.nexus.disabled to false as a system property?

I tried it, but unfortunately, this parameter did not have any effect.


**I am now sure that this is definitely not a ByteBuddy issue.**

I decompiled the bytecode information enhanced by ByteBuddy and then got rid of ByteBuddy and only used ASM to hardcode these bytecode into the target class. The parameter `-XX:+TraceClassLoading` confirms that none of ByteBuddy's classes are loaded during the process.

I even removed the `Class<MyURLClassLoader>` element from the `classes` attribute of the `SystemClassLoader` object through reflection after performing the detach operation, and the heap dump showed that the number of class loader instances was 0.
However, even so, the JVM does not want to unload it as long as I use a version lower than Java 11. Maybe only the JVM team knows what's really going on  :shrug:.
